src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -401,6 +504,269 @@ private static void AssertNoImplicitInterpolatedStringHandlerConversions(Immutab
+                {
+                    Debug.Assert(previousUnmergedAppendLiteralCall is null);
+                    (previousUnmergedAppendLiteralCall, previousUnmergedStringLiteral) = (call, literal);
+                }
+                else
+                {
+                    // already has pending AppendLiteral; merge into its parameter string
+                    mergePreviousAppendLiteralStringIntoBuilder();
+                    currentBeingMergedStringConst.Builder.Append(value);
+                }
+                return true;
+
+            }
+
+            (ImmutableArray<BoundExpression> Calls, int IncreasedLiteralLength, int FilledHolesCount) shouldBeAsIs()
+            {
+                result.Free();
+                currentBeingMergedStringConst.Free();
+                return (parts, 0, 0);
+            }
+
+            static MethodSymbol? tryGetAppendLiteralCallFromReceiver(BoundExpression receiver)
+            {
+                Debug.Assert(receiver.Type is not null);
+                foreach (var candidate in receiver.Type!.GetMembers(BoundInterpolatedString.AppendLiteralMethod))
+                {
+                    if (candidate is not MethodSymbol methodCandidate)
+                        continue;
+                    if (methodCandidate.IsStatic)
+                        continue;
+                    if (methodCandidate is not { Parameters: [{ Type.SpecialType: SpecialType.System_String }] })
+                        continue;
+                    if (methodCandidate.DeclaredAccessibility != Accessibility.Public)
+                        continue;
+                    return methodCandidate;
+
+                }
+                return null;
+            }
+
+            // This must be called just before modifying previous... or currentBeingMergedStringConst
+            void mergePreviousAppendLiteralStringIntoBuilder()
+            {
+                // other than null or empty literal
+                if (previousUnmergedStringLiteral is { ConstantValueOpt.StringValue: { Length: not 0 } value })
+                {
+                    currentBeingMergedStringConst.Builder.Append(value);
+                }
+                // We will have to newly generate a AppendLiteral without reusing the existing one
+                previousUnmergedAppendLiteralCall = null;
+                previousUnmergedStringLiteral = null;
+            }
+
+            bool tryAddPendingAppendLiteralCallToResult(BoundExpression? handler)
+            {
+                if (previousUnmergedAppendLiteralCall is not null)
+                {
+                    Debug.Assert(currentBeingMergedStringConst.Length == 0);
+                    // single append call; add it as is
+                    result.Add(previousUnmergedAppendLiteralCall);
+                }
+                else
+                {
+                    BoundLiteral? literal = null;
```

Comment:
I wonder if we should temporarily retain not only `BoundLiteral` but also other `BoundExpression` that has string `ConstantValueOpt`.
Those other than `BoundLiteral` are not ideal but fine as arguments of `AppendLiteral`.